File: 在线仓库/modelzoo-master/samples/built-in/count/CrowdCount/script/crowdcount_preprocess.py
# Copyright (c) ModelZoo. 2025-2025. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import cv2
import numpy as np
import argparse
import json
from tqdm import tqdm


def preprocess_image(img, target_size=(224, 224)):
    """预处理图像：RGB转BGR、调整尺寸、归一化到[0, 1]"""
    # 调整尺寸
    img_resized = cv2.resize(img, target_size, interpolation=cv2.INTER_LINEAR)
    # 归一化到[0, 1]
    img_norm = img_resized.astype(np.float32) / 255.0
    # 转换为NCHW格式 (1, 3, H, W)
    img_nchw = np.expand_dims(img_norm.transpose(2, 0, 1), axis=0)
    return np.ascontiguousarray(img_nchw)


def image_to_bin(image_path, output_bin_path, target_size=(224, 224)):
    """将单张图像转换为预处理后的bin文件"""
    try:
        img = cv2.imread(image_path)
    except Exception as e:
        print(f"警告: 无法读取图像 {image_path}: {e}")
        return False

    # 预处理
    try:
        processed = preprocess_image(img, target_size)
        processed.tofile(output_bin_path)
        return True
    except Exception as e:
        print(f"警告: 处理图像 {image_path} 失败: {e}")
        return False


def process_images(file_list_path, output_dir, bin_list_file, target_size=(224, 224), max_count=5000):
    """批量处理图像并生成bin文件列表"""
    os.makedirs(output_dir, exist_ok=True)

    # 读取文件列表
    if not os.path.exists(file_list_path):
        print(f"错误: 文件列表 {file_list_path} 不存在")
        return

    with open(file_list_path, 'r') as f:
        try:
            data = json.load(f)
            image_files = [item[0] for item in data.get('fileList', [])]
        except json.JSONDecodeError as e:
            print(f"错误: 解析JSON文件失败: {e}")
            return

    # 限制处理数量
    process_files = image_files[:max_count]

    # 处理图像
    with open(bin_list_file, 'w') as f_list:
        for filename in tqdm(process_files, desc="预处理进度"):
            if not os.path.exists(filename):
                print(f"警告: 文件不存在，跳过: {filename}")
                continue

            # 生成输出路径
            base_name = os.path.splitext(os.path.basename(filename))[0]
            output_path = os.path.join(output_dir, f"{base_name}.bin")

            # 转换为bin文件
            if image_to_bin(filename, output_path, target_size):
                f_list.write(f"{os.path.abspath(output_path)}\n")


def main():
    parser = argparse.ArgumentParser(description='CrowdCount图像预处理脚本')
    parser.add_argument('--file_list', default='../data/file_list.json',
                        help='包含图像路径的JSON文件路径')
    parser.add_argument('--output_dir', default='../data/preprocess/bin',
                        help='预处理后bin文件输出目录')
    parser.add_argument('--target_size', type=int, nargs=2, default=[224, 224],
                        help='模型输入尺寸，格式为 宽 高，默认224 224')
    parser.add_argument('--max_count', type=int, default=5000,
                        help='最大处理图像数量，默认5000张')

    args = parser.parse_args()

    # 生成bin文件列表路径
    bin_list_file = os.path.join(os.path.dirname(args.output_dir), 'file_list.txt')

    # 执行预处理
    process_images(
        file_list_path=args.file_list,
        output_dir=args.output_dir,
        bin_list_file=bin_list_file,
        target_size=tuple(args.target_size),
        max_count=args.max_count
    )
    print(f"预处理完成，结果保存至: {args.output_dir}")
    print(f"文件列表保存至: {bin_list_file}")


if __name__ == "__main__":
    main()

File: 在线仓库/modelzoo-master/samples/common/infer/preprocess/crowdcount_preprocess.cpp
/*
 * Copyright (c) ModelZoo. 2025-2025. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "crowdcount_preprocess.h"
#include "log.h"
#include "utils.h"
#include <opencv2/opencv.hpp>
#include <fstream>
#include <vector>
#include <string>
#include <sys/stat.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

namespace Infer {
namespace CrowdCountNS {

constexpr int DEFAULT_TARGET_WIDTH = 224;
constexpr int DEFAULT_TARGET_HEIGHT = 224;
constexpr int BYTE_BIT_NUM = 8;
constexpr float PIXEL_MAX_VALUE = 255.0f;
constexpr int CHANNEL_NUM = 3;
constexpr int BATCH_SIZE = 1;

static cv::Mat PreprocessImage(const cv::Mat& img, int targetWidth, int targetHeight) {
    // 调整尺寸
    cv::Mat imgResized;
    cv::resize(img, imgResized, cv::Size(targetWidth, targetHeight), 0, 0, cv::INTER_LINEAR);

    // 归一化到[0, 1]
    cv::Mat imgNorm;
    imgResized.convertTo(imgNorm, CV_32FC3, 1.0f / PIXEL_MAX_VALUE);

    return imgNorm;
}

static cv::Mat HwcToChw(const cv::Mat& hwcImg) {
    CV_Assert(!hwcImg.empty() && hwcImg.channels() == CHANNEL_NUM);
    std::vector<cv::Mat> channels;
    cv::split(hwcImg, channels);

    cv::Mat chwMat(CHANNEL_NUM, hwcImg.rows * hwcImg.cols, CV_32F);
    chwMat = cv::Scalar(0);

    for (int channelIdx = 0; channelIdx < CHANNEL_NUM; ++channelIdx) {
        cv::Mat flatChannel = channels[channelIdx].reshape(1, 1).clone();
        flatChannel.convertTo(flatChannel, CV_32F);
        flatChannel.copyTo(chwMat.row(channelIdx));
    }

    return chwMat;
}

    static Result ReadImageToTensorBuf(const cv::Mat& chwImg, TensorDesc& desc, TensorBuf& inBuf) {
        LOG(INFO) << "ReadImgFileToBuf: desc.dimCount " << desc.dimCount;
        const float *srcData = chwImg.ptr<float>();
        // dlite核，不需要处理字节对齐，直接读取数据
        if (desc.defaultStride == 0) {
            memcpy(static_cast<float *>(inBuf.GetRawPtr()), srcData, desc.defaultSize);
            return SUCCESS;
        }

        // dpico核，需要处理字节对齐
        int64_t loopTimes = 1;
        for (size_t loop = 0; loop < desc.dimCount - 1; loop++) {
            loopTimes *= desc.dims[loop];
        }

        int64_t width = desc.dims[desc.dimCount - 1]; /* dims last dim is width */
        size_t dataSize = desc.typeSize / BYTE_BIT_NUM;
        size_t strideElemNum = inBuf.stride / dataSize;

        for (int64_t loop = 0; loop < loopTimes; loop++) {
            float *destPtr = static_cast<float *>(inBuf.GetRawPtr()) + loop * strideElemNum;
            const float *srcPtr = srcData + loop * width;

            if (destPtr == nullptr || srcPtr == nullptr) {
                LOG(ERROR) << "Null pointer detected";
                return FAILED;
            }

            memcpy(destPtr, srcPtr, width * dataSize);
        }
        return SUCCESS;
    }

bool CrowdCountPreprocess(std::vector<std::string>& fileList,
                         std::vector<TensorBuf>& tensorBufs,
                         std::vector<TensorDesc>& tensorDescs) {
    auto cfg = ReadCfgFile("../data/cfg.txt");
    std::string saveBin = cfg["save_preprocess_bin"];

    if (!saveBin.empty() && !CreateDirectoryRecursive(saveBin)) {
        LOG(ERROR) << "Create preprocess bin dir failed: " << saveBin;
        return false;
    }

    LOG(INFO) << "Processing " << fileList.size() << " images...";

    for (size_t i = 0; i < fileList.size(); ++i) {
        std::string imgPath = fileList[i];
        LOG(INFO) << "Processing image: " << imgPath;

        // 提取文件名（不含扩展名）
        size_t start = imgPath.find_last_of("/");
        size_t end = imgPath.find_last_of(".");
        std::string outputName = imgPath.substr(start + 1, end - start - 1);

        // 读取图像并转换为RGB
        cv::Mat img = cv::imread(imgPath);
        if (img.empty()) {
            LOG(ERROR) << "Read image failed: " << imgPath;
            continue;
        }

        // 预处理图像
        cv::Mat processedImg = PreprocessImage(img, DEFAULT_TARGET_WIDTH, DEFAULT_TARGET_HEIGHT);

        // 转换为CHW格式
        cv::Mat chwImg = HwcToChw(processedImg);

        if (!chwImg.isContinuous()) {
            chwImg = chwImg.clone();
        }

        // 填充tensor缓冲区
        if (ReadImageToTensorBuf(chwImg, tensorDescs[i], tensorBufs[i]) != SUCCESS) {
            LOG(ERROR) << "Failed to fill tensor buffer for: " << imgPath;
            continue;
        }

        // 保存预处理结果
        if (!saveBin.empty()) {
            std::string binPath = saveBin + "/" + outputName + ".bin";
            std::ofstream binFile(binPath, std::ios::binary);
            if (binFile.is_open()) {
                size_t dataSize = DEFAULT_TARGET_WIDTH * DEFAULT_TARGET_HEIGHT * CHANNEL_NUM * sizeof(float);
                binFile.write(reinterpret_cast<char*>(chwImg.data), dataSize);
                binFile.close();
                if (chmod(binPath.c_str(), 0755) != 0) {
                    LOG(WARNING) << "Failed to set permissions for: " << binPath;
                }
                LOG(INFO) << "Saved preprocess bin: " << binPath;
            } else {
                LOG(WARNING) << "Failed to open bin file: " << binPath;
            }
        }
    }

    LOG(INFO) << "PreProcessing completed successfully!";
    return true;
}

}  // namespace CrowdCountNS
}  // namespace Infer